plat/kvm/x86/acpi.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Authors: Cristian Vijelie <cristianvijelie@gmail.com>
+ *
+ * Copyright (c) 2021, University Politehnica of Bucharest. All rights reserved.
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <uk/print.h>
+#include <string.h>
+#include <x86/acpi/acpi.h>
+#include <x86/acpi/madt.h>
+
+__u8 acpi_version;
+struct RSDPDescriptor *rsdp;
+struct RSDPDescriptor20 *rsdp2;
+struct RSDT *rsdt;
+struct XSDT *xsdt;
+struct MADT *madt;
+
+static inline __u8 detect_acpi_version(__u64 *ptr)
+{
+	struct RSDPDescriptor *rsdp = (struct RSDPDescriptor *)ptr;
+
+	return rsdp->Revision;
+}
+
+static inline __s8 rsdp_checksum(__u64 *ptr)
+{
+	__u32 checksum = 0;
+	__u8 *i;
+
+	for (i = (__u8 *)ptr; i < (__u8 *)ptr + sizeof(struct RSDPDescriptor);
+	     i++)
+		checksum += *i;
+
+	return (__u8)checksum == 0 ? 0 : -1;
+}
+
+/*
+ * Not yet implemented, because I have not yet detected ACPI version 2 or
+ * greater.
+ */
+
+static inline __s8 rsdp2_checksum(__u64 *ptr)
+{
+	return 0;
+}
+
+/*
+ * Do checksum for any ACPI table, except RSDP.
+ */
+
+static inline __s8 acpi_checksum(struct ACPISDTHeader *h)
+{
+	__u8 checksum = 0;
+	__u32 i;
+
+	for (i = 0; i < h->Length; i++)
+		checksum += ((__u8 *)h)[i];
+
+	return checksum == 0 ? 0 : -1;
+}
+
+/*
+ * Find the ACPI Root System Descriptor Table (RSDT), and check if it's valid.
+ * For this, the Root System Descriptor Pointer (RSDP) must be found, in the
+ * physical memory area 0xe0000 -> 0xfffff.
+ */
+
+void find_rsdt(void)
+{
+	__u64 start_addr = 0xe0000, end_addr = 0xfffff;
+	__u64 i;
+	__u64 *ptr;
+
+	for (i = start_addr; i < end_addr; i += 16) {
+		ptr = (__u64 *)i;
+		if (!memcmp(ptr, "RSD PTR ", 8)) {
+			uk_pr_info("Found RSDP at %p\n", ptr);
+			break;
+		}
+	}
+
+	uk_pr_info("ACPI version detected: ");
+	if (!detect_acpi_version(ptr)) {
+		uk_pr_info("1.0\n");
+		if (!rsdp_checksum(ptr)) {
+			uk_pr_debug("RSDP is valid\n");
+			acpi_version = 1;
+			rsdp = (struct RSDPDescriptor *)ptr;
+			if (!acpi_checksum(
+				(struct ACPISDTHeader *)rsdp->RsdtAddress)) {
+				rsdt = (struct RSDT *)rsdp->RsdtAddress;
+				uk_pr_info("RSDT is valid and present at %p\n",
+					   rsdt);
+			} else {
+				uk_pr_err("RSDT is not valid\n");
+				return;
+			}
+		} else {
+			uk_pr_err("RSDP is not valid\n");
```

Comment:
This looks to me that you could flatten the nested `if` statements when you exit (or `goto` to an error exit label) when you check for failing cases, e.g.:
```  c
if (rsdp_checksum(ptr) != 0)
   uk_pr_err("Invalid RSDP\n");   
   return;
}

acpi_version = 1;
rsdp = (struct RSDPDescriptor *) ptr;
rsdt = (struct RSDT *) rsdp->RsdtAddress;

if (acpi_checksum(rsdt) != 0) {
   uk_pr_err("Invalid RSDT\n");
   return;
}

uk_pr_info("RSDT present at %p\n", rsdt);
```